pkg/scheduler/internal/queue/pod_backoff.go

```diff
@@ -0,0 +1,119 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"sync"
+	"time"
+
+	ktypes "k8s.io/apimachinery/pkg/types"
+)
+
+// PodBackoffMap is a structure that stores backoff related information for pods
+type PodBackoffMap struct {
+	// lock for performing actions on this PodBackoffMap
+	lock sync.Mutex
+	// initial backoff duration
+	initialDuration time.Duration
+	// maximal backoff duration
+	maxDuration time.Duration
+	// map for pod -> number of attempts for this pod
+	podAttempts map[ktypes.NamespacedName]int
+	// map for pod -> lastUpdateTime pod of this pod
+	podLastUpdateTime map[ktypes.NamespacedName]time.Time
+}
+
+// NewPodBackoffMap creates a PodBackoffMap with initial duration and max duration.
+func NewPodBackoffMap(initialDuration, maxDuration time.Duration) *PodBackoffMap {
+	return &PodBackoffMap{
+		initialDuration:   initialDuration,
+		maxDuration:       maxDuration,
+		podAttempts:       make(map[ktypes.NamespacedName]int),
+		podLastUpdateTime: make(map[ktypes.NamespacedName]time.Time),
+	}
+}
+
+// GetBackoffTime returns the time that nsPod completes backoff
+func (pbm *PodBackoffMap) GetBackoffTime(nsPod ktypes.NamespacedName) (time.Time, bool) {
+	pbm.lock.Lock()
+	defer pbm.lock.Unlock()
+	if _, found := pbm.podAttempts[nsPod]; found == false {
+		return time.Time{}, false
+	}
+	lastUpdateTime := pbm.podLastUpdateTime[nsPod]
+	backoffDuration := pbm.calculateBackoffDuration(nsPod)
+	backoffTime := lastUpdateTime.Add(backoffDuration)
+	return backoffTime, true
+}
+
+// TryBackoffAndWait tries to perform backoff for a non-preempting pod.
+// it is invoked from factory.go if util.PodPriorityEnabled() returns false.
+func (pbm *PodBackoffMap) TryBackoffAndWait(nsPod ktypes.NamespacedName, stop <-chan struct{}) bool {
+	pbm.lock.Lock()
+	defer pbm.lock.Unlock()
+	backoffDuration := pbm.calculateBackoffDuration(nsPod)
+	select {
+	case <-time.After(backoffDuration):
+		return true
+	case <-stop:
+		return false
+	}
+}
+
+// calculateBackoffDuration is a helper function for calculating the backoffDuration
+// based on the number of attempts the pod has made.
+func (pbm *PodBackoffMap) calculateBackoffDuration(nsPod ktypes.NamespacedName) time.Duration {
+	backoffDuration := pbm.initialDuration
+	if _, found := pbm.podAttempts[nsPod]; found {
+		for i := 1; i < pbm.podAttempts[nsPod]; i++ {
+			backoffDuration = backoffDuration * 2
+			if backoffDuration > pbm.maxDuration {
+				return pbm.maxDuration
+			}
+		}
+	}
+	return backoffDuration
+}
+
+// clearPodBackoff removes all tracking information for nsPod.
+// Lock is supposed to be acquired by caller.
+func (pbm *PodBackoffMap) clearPodBackoff(nsPod ktypes.NamespacedName) {
+	delete(pbm.podAttempts, nsPod)
```

Comment:
Many of the functions here are thread safe, but this one is not. It is left to the caller to acquire the lock. I guess the reason is that `CleanupPodsCompletesBackingoff` uses this function. Please create a thread safe version of this function which is public (starts with a capital letter) and use the new function in other modules.